### PR TITLE
Remove ocamlformat from the linter and the pre-commit hook.

### PR DESCRIPTION
**Pull request by ppedrot**
At last I'll be able to rebase some old stuff that was unfortunate enough to touch funind.

### CODE REVIEW

**Review by Matafou — APPROVED** [review]
No objection.

**Review by maximedenes — APPROVED** [review]

**Comment by Zimmi48** [comment]
Who's assigning?

**Comment by Matafou** [comment]
I can. Can I use coqbot instead of merge.sh?

**Comment by Zimmi48** [comment]
Yep :wink:

**Review by ejgallego — DISMISSED** [review]
I am OK making `ocamlformat` optional, however I don't follow the premise that motivates this PR

> At last I will be able to rebase

I am not sure how the current infrastructure does create problems with rebase.

Indeed that takes us to my worry about this PR, it removes all ocamlformat configuration from Coq, however the hook in CI and pre-commit remain there, untested. It seems to me that this needs discussion on both accounts.

**Comment by ppedrot** [comment]
> I am not sure how the current infrastructure does create problems with rebase.

This has been discussed at large already. Rebasing can change the output of ocamlformat on an abitrary number of lines of codes, leading to the subsequent patches not to apply anymore. Ironically, funind is particularly prone to this problem because it defines huge blocks of code that tend to get rewritten a lot by auto-formatting.

**Comment by ejgallego** [comment]
> This has been discussed at large already. Rebasing can change the output of ocamlformat on an abitrary number of lines of codes, leading to the subsequent patches not to apply anymore.

So what creates your problem is the git hook, not the files you are updating here ? I still think the message is misleading, in the sense that this change won't help any rebase w.r.t. previous trees, unless I'm missing something.

[p.s: you could try https://github.com/hhugo/merge-fmt on your old tree and see if that helps with the pain]

**Comment by ppedrot** [comment]
AFAIR the hook doesn't apply on a rebase, the problem is that the linter checks *all* commits in the PR. So you have to forcefully apply the formatter on each commit of the rebase, leading to the above problem.

**Comment by ejgallego** [comment]
> the problem is that the linter checks _all_ commits in the PR. So you have to forcefully apply the formatter on each commit of the rebase, leading to the above problem.

Is that true? Checking at the code it doesn't seem the case.

**Comment by ppedrot** [comment]
Right, so it's the probably the opposite: the linter checks the last commit but the rebase applies the formatter. In either case I remember having to perform impossible rebases.

**Comment by ejgallego** [comment]
Indeed the hook may create problems if you don't have merge-fmt hooked into git ; anyways it makes no sense to merge this PR and leave the infrastructure in place so discussion is needed to see what we do.

Meanwhile not having this in the tree is not a blocker as just uninstalling ocamlformat should avoid that problems and we can always override the linter on merge.

**Comment by gares** [comment]
> I am OK making `ocamlformat` optional

Can we just move the linter job to allowed failure?

**Comment by Zimmi48** [comment]
If we do bypass a linter failure due to formatting, then it is going to fail every time it runs and therefore it will lose all interest, including the rest of the linting.

**Comment by gares** [comment]
So, am I right in saying that this PR should be amended by also removing all CI infrastructure for code formatting? (or closed)

**Comment by Zimmi48** [comment]
Well... if we are sure that we definitely want to abandon `ocamlformat`, then sure. However, if we might want to reintroduce it on some other part of the code base, then I'd rather keep the infrastructure in place (that's why I didn't say anything related to this infrastructure in the first place). It's not going to hurt anyone in the meantime.

**Comment by gares** [comment]
Hum, then maybe we could add a .ocamlformat for a file that is not subject to many changes and rebase, for example configure.ml, just to keep the CI infrastructure tested as per @ejgallego request, and remove the ones for funind and micromega as this PR does.

**Comment by Zimmi48** [comment]
I would be happy with this solution.

**Comment by ejgallego** [comment]
> I would be happy with this solution.

I am OK if we wanna go this route.

I still do think tho that the current rules for ocamlformat are a bi strict on our side, most other projects don't enforce formatting in CI or take care of it fully transparently. That means some manual application of `make lint` after merge, but IMHO it is less intrusive than the current setup, even tho if I think Pierre-Marie fobia to this tooling is a tad irrational :) , but well, I guess quite a bit of people have huge nightmares with [insert your own tooling].

I think that in general keeping plugins such as funind or micromega neat should still be beneficial, so I'd rather remove the hooks and still suggest people do `make fmt` when they are happy, than having a hook for a file we don't care.

**Comment by ppedrot** [comment]
> I'd rather remove the hooks and still suggest people do make fmt

All right, I've added a patch to do that. It's easy to revert if we want to retrieve that code.

> keeping plugins such as funind or micromega neat should still be beneficial

I am possibly a victim of an irrational phobia regarding ocamlformat, but I also think that it's a bit delusional that the major problem with funind is code layout and that it can be solved automagically with a formatting tool. Truth is, funind is horrible at so many levels I do not even want to discuss it here again. Funind delenda est, equationibusque intercambianda.

**Comment by gares** [comment]
> equationibusque

At some point I've heard that equations, which is great, cannot replace all funind does (as in will not). Was I dreaming? Should we stop believing in this thing?

**Review by gares — APPROVED** [review]

File: dev/doc/changes.md [hunk 1]
```diff
@@ -4,6 +4,12 @@
   comes from a notation. Use `None` if not and `Some foo` to tell to
   print such TacGeneric surrounded with `foo:( )`.
 
+### Code formatting
+
+- The automatic code formatting tool `ocamlformat` has been disabled and its
+  git hook removed. If desired, automatic formatting can be achieved by calling
+  the `fmt` target of the dune build system.
```
Comment (Zimmi48) [inline dev/doc/changes.md:11]: This claim is inconsistent with the removal of the `.ocamlformat` files that is still part of this PR. With this removal, this target will do nothing.

**Review by Zimmi48 — CHANGES_REQUESTED** [review]
The current status does not make any sense. Either you keep the infrastructure in the linter but disable ocamlformat for micromega and funind (in which case you could activate it on a simple file like `configure.ml` so that it continues to be tested as was proposed by @gares) or you remove the linter / pre-commit infrastructure but keep the `.ocamlformat` files in micromega and funind (in this case the PR should be retitled) or you remove all traces of ocamlformat entirely from the code base and you don't claim that it can still be used if desired.

**Comment by ppedrot** [comment]
@Zimmi48 I've reverted the changes to the ocamlformat files.

**Review by Zimmi48 — APPROVED** [review]

**Comment by ppedrot** [comment]
@Matafou AFAICT this is ready.

**Comment by ppedrot** [comment]
@Matafou Can we get this merged? I don't want to see another diff that is 3 times as big as needed as in #13021 again ever.

**Comment by gares** [comment]
If @matafou is not available I can take over.
To my understanding the request for change by @ejgallego was addressed.
If I head nothing I'll merge tomorrow.

**Comment by gares** [comment]
I got a personal communication by @Matafou to take over this PR.

**Comment by gares** [comment]
@coqbot merge now

**Pull request merged**